src/crypto/test/crypto.cpp

```diff
@@ -375,15 +384,120 @@ static const vector<uint8_t>& getRawKey()
+  k->encrypt(h.get_iv(), p, nullb, p.p, h.tag);
 
-  KeyAesGcm k2(getRawKey());
-  REQUIRE(k2.decrypt(h.get_iv(), h.tag, p, nullb, p.p));
+  auto k2 = crypto::make_key_aes_gcm(getRawKey());
+  REQUIRE(k2->decrypt(h.get_iv(), h.tag, p, nullb, p.p));
+}
+
+TEST_CASE("AES mbedTLS vs OpenSSL")
+{
+  auto key = getRawKey();
+
+  // std::cout << "Contents: ";
+  // for (size_t i = 0; i < contents_.size(); i++)
+  //   printf("%02x", contents_[i]);
+  // std::cout << std::endl;
+
+  GcmHeader<1234> h;
+
+  { // mbedTLS -> OpenSSL
+    auto mbed = std::make_unique<KeyAesGcm_mbedTLS>(key);
+    auto ossl = std::make_unique<KeyAesGcm_OpenSSL>(key);
+
+    std::vector<uint8_t> encrypted(contents_.size());
+    mbed->encrypt(h.get_iv(), contents_, nullb, encrypted.data(), h.tag);
+
+    std::vector<unsigned char> rawP(contents_.size(), 'x');
+    Buffer p{rawP.data(), rawP.size()};
+
+    std::vector<uint8_t> decrypted(contents_.size());
+    REQUIRE(ossl->decrypt(
+      h.get_iv(),
+      h.tag,
+      {encrypted.data(), encrypted.size()},
+      nullb,
+      decrypted.data()));
+
+    REQUIRE(decrypted.size() == contents_.size());
+    REQUIRE(memcmp(decrypted.data(), contents_.data(), sizeof(contents_)) == 0);
+  }
+
+  { // OpenSSL -> mbedTLS
+    auto mbed = std::make_unique<KeyAesGcm_mbedTLS>(key);
+    auto ossl = std::make_unique<KeyAesGcm_OpenSSL>(key);
+
+    std::vector<uint8_t> encrypted(contents_.size());
+    ossl->encrypt(h.get_iv(), contents_, nullb, encrypted.data(), h.tag);
+
+    std::vector<unsigned char> rawP(contents_.size(), 'x');
+    Buffer p{rawP.data(), rawP.size()};
+
+    std::vector<uint8_t> decrypted(contents_.size());
+    CBuffer encbuf{encrypted.data(), encrypted.size()};
+    REQUIRE(mbed->decrypt(h.get_iv(), h.tag, encbuf, nullb, decrypted.data()));
+
+    REQUIRE(decrypted.size() == contents_.size());
+    REQUIRE(memcmp(decrypted.data(), contents_.data(), sizeof(contents_)) == 0);
+  }
+}
+
+TEST_CASE("AES mbedTLS vs OpenSSL + AAD")
+{
+  auto key = getRawKey();
+
+  // std::cout << "Contents: ";
+  // for (size_t i = 0; i < contents_.size(); i++)
+  //   printf("%02x", contents_[i]);
+  // std::cout << std::endl;
+
+  GcmHeader<1234> h;
+  std::vector<uint8_t> aad(123, 'y');
+
+  { // mbedTLS -> OpenSSL
+    auto mbed = std::make_unique<KeyAesGcm_mbedTLS>(key);
+    auto ossl = std::make_unique<KeyAesGcm_OpenSSL>(key);
+
+    std::vector<uint8_t> encrypted(contents_.size());
+    mbed->encrypt(h.get_iv(), contents_, aad, encrypted.data(), h.tag);
+
+    std::vector<unsigned char> rawP(contents_.size(), 'x');
+    Buffer p{rawP.data(), rawP.size()};
+
+    std::vector<uint8_t> decrypted(contents_.size());
+    REQUIRE(ossl->decrypt(
+      h.get_iv(),
+      h.tag,
+      {encrypted.data(), encrypted.size()},
+      aad,
+      decrypted.data()));
+
+    REQUIRE(decrypted.size() == contents_.size());
+    REQUIRE(memcmp(decrypted.data(), contents_.data(), sizeof(contents_)) == 0);
+  }
+
+  { // OpenSSL -> mbedTLS
```

Comment:
Doctest has a nice INFO() macro we often use, which if you run with -s will get logged to stdout. Here it'd be nice to have INFO("OpenSSL vs mbedTLS") or something like that rather than a comment.